lib/job/arBaseJob.class.php

```diff
@@ -284,34 +284,94 @@ private function createJobsDownloadsDirectory()
    * Create ZIP file from results
    *
    * @param string  Path of file to write CSV data to
+   * @param string  Optional array of digital object ids to include
    *
    * @return int  success bool
    */
-  protected function createZipForDownload($path)
+  protected function createZipForDownload($path, $digitalObjects = array())
   {
+    $errors = array();
+
     if (!is_writable($this->getJobsDownloadDirectory()))
     {
-      return false;
+      $errors[] = $this->i18n->__('Cannot write to directory');
     }
-
-    $zip = new ZipArchive();
-
-    $success = $zip->open($this->getDownloadFilePath(), ZipArchive::CREATE | ZipArchive::OVERWRITE);
-
-    if ($success == true)
+    else
     {
-      foreach(scandir($path) as $file)
+      
+      $zip = new ZipArchive();
+
+      if (true != $zip->open($this->getDownloadFilePath(), ZipArchive::CREATE | ZipArchive::OVERWRITE))
+      {
+        $errors[] = $this->i18n->__('Cannot initialize file');
+      }
+      else 
       {
-        if (!is_dir($file))
+        // Check if there are any digital objects to include and if yes, add them to the zip
+        if (!empty($digitalObjects))
         {
-          $zip->addFile($path . DIRECTORY_SEPARATOR . $file, $file);
+          // Keep track of file names so we can append a bracketed number if any are duplicated]
+          $fileNames = array();
+          // Iterate digital object ids
+          foreach($digitalObjects as $id)
+          {
+            $do = QubitDigitalObject::getById($id);
+            if (null != $do)
+            {
+              $doPath = $do->getAbsolutePath();
+              if (file_exists($doPath))
+              {
+                $fileName = basename($doPath);
+                if (!array_key_exists($fileName, $fileNames))
+                {
+                  // filename not used yet - add to tracker
+                  $fileNames[$fileName] = 0;
+                }
+                else
+                {
+                  // filename has been used - increment counter and add brackets
+                  $fileNames[$fileName]++;
+                  $doPathInfo = pathinfo( $doPath );
+                  $fileName = $doPathInfo['filename'] . ' (' . $fileNames[$fileName] . ').' . $doPathInfo['extension'];
+                }
+                try
+                {
+                  $zip->addFile($doPath, $fileName);
+                }
+                catch (Exception $e)
+                {
+                  if ($this->user->isAdministrator())
+                  {
+                    $errors[] = ('Exception: '.$e->getMessage());
+                  }
```

Comment:
Why the administrator check here and not in the previous two possible errors?

Reply:
Have expanded this in revised code. Administrators should see full Exception (message), but any other user will see some more helpful text. To be reviewed.